Add Ninja configuration to Linux and Windows CI

We should add a configuration to CI for building with the Ninja generator on Linux and Windows.

This will test not only that Ninja works, but also that parallel builds still work (since Ninja always runs in parallel).

Add an appropriate number of stages in Jenkins CI for each platform to use the Ninja build engine:
- [ ] Linux
- [x] Windows (once #1272 is merged) (Completed with Anita's PR #1415 )

Thanks @ionutbalutoiu . Seems like it built fine on one system/configuration and had 2 test issues but failed prematurely on the other (Debug Cross-compile config/machine).
A couple of questions - 1) Did you install Visual Studio 2017 or only the Build Tools?
2) On the debug Cross-compile build failure, the following error is reported.
Can you follow the instruction in the document https://github.com/Microsoft/openenclave/blob/master/docs/GettingStartedDocs/GettingStarted.Windows.md and verify that all the correct components are there. This points to issues with clang. Open a command window and do a where clang. It should report C:\Program Files\LLVM\bin\clang.exe. Please read the online guide to verify other components. Thanks.

-- The C compiler identification is Clang 7.0.1
CMake Error at C:/Program Files/CMake/share/cmake-.13/Modules/CMakeDetermineCompilerId.cmake:802 (message):
  The Clang compiler tool
    "C:/Program Files/LLVM/bin/clang-7.exe"
  targets the MSVC ABI but has a GNU-like command-line interface.  This is
  not supported.  Use 'clang-cl' instead, e.g.  by setting 'CC=clang-cl' in
  the environment.
Call Stack (most recent call first):
  C:/Program Files/CMake/share/cmake-3.13/Modules/CMakeDetermineCCompiler.cmake:113 (CMAKE_DIAGNOSE_UNSUPPORTED_CLANG)
  CMakeLists.txt:28 (project)

--- a/.jenkins/Jenkinsfile
+++ b/.jenkins/Jenkinsfile
@@ -14,8 +14,8 @@ def ACCTest(String label, String compiler, String build_type) {
                     }
                     withEnv(["CC=${c_compiler}","CXX=${cpp_compiler}"]) {
                         sh """
-                        cmake ${WORKSPACE} -DCMAKE_BUILD_TYPE=${build_type}
-                        make
+                        cmake ${WORKSPACE} -G Ninja -DCMAKE_BUILD_TYPE=${build_type}
+                        ninja -v
                         ctest --output-on-failure
                         """
                     }
@@ -42,8 +42,8 @@ def simulationTest(String version, String platform_mode, String build_type) {
                     dir('build') {
                         withEnv(["CC=clang-7","CXX=clang++-7","OE_SIMULATION=1"]) {
                             sh """
-                            cmake ${WORKSPACE} -DCMAKE_BUILD_TYPE=${build_type} -DUSE_LIBSGX=${use_libsgx}
-                            make
+                            cmake ${WORKSPACE} -G Ninja -DCMAKE_BUILD_TYPE=${build_type} -DUSE_LIBSGX=${use_libsgx}
+                            ninja -v
                             ctest --output-on-failure
                             """
                         }
@@ -66,8 +66,8 @@ def ACCContainerTest(String label, String version) {
                     dir('build') {
                         withEnv(["CC=clang-7","CXX=clang++-7"]) {
                             sh """
-                            cmake ${WORKSPACE} -DCMAKE_BUILD_TYPE=RelWithDebInfo
-                            make
+                            cmake ${WORKSPACE} -G Ninja -DCMAKE_BUILD_TYPE=RelWithDebInfo
+                            ninja -v
                             ctest --output-on-failure
                             """
                         }
@@ -107,8 +107,8 @@ def checkDevFlows(String version) {
 
                     dir('build') {
                         sh '''
-                        cmake ${WORKSPACE} -DUSE_LIBSGX=OFF
-                        make
+                        cmake ${WORKSPACE} -G Ninja -DUSE_LIBSGX=OFF
+                        ninja -v
                         '''
                         // Note that `make package` is not expected to work
                         // without extra configuration.
@@ -130,8 +130,8 @@ def win2016LinuxElfBuild(String version, String build_type) {
                     dir('build') {
                         withEnv(["CC=clang-7","CXX=clang++-7"]) {
                             sh """
-                            cmake ${WORKSPACE} -DCMAKE_BUILD_TYPE=${build_type} -DUSE_DEBUG_MALLOC=OFF
-                            make
+                            cmake ${WORKSPACE} -G Ninja -DCMAKE_BUILD_TYPE=${build_type} -DUSE_DEBUG_MALLOC=OFF
+                            ninja -v
                             """
                         }
                     }

--- a/scripts/ansible/roles/linux/openenclave/tasks/environment-setup.yml
+++ b/scripts/ansible/roles/linux/openenclave/tasks/environment-setup.yml
@@ -26,6 +26,7 @@
       - "gcc"
       - "gdb"
       - "g++"
+      - "ninja-build"
       - "apt-transport-https"
       - "autoconf"
       - "doxygen"

